From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stephen Chenney <schenney@chromium.org>
Date: Thu, 3 Sep 2026 17:41:55 -0700
Subject: [M152] [HiC] Fix privacy leaks due to feature flag checks

Original change's description:
> [HiC] Fix privacy leaks due to feature flag checks
>
> When a frame that does not enable HTML-in-Canvas embeds a same-origin
> iframe that does enable the origin trial (through headers in the
> subframe request) the sub-frame can use drawElementImage with
> cross-origin filter content and the checks in the main frame painting
> code will not remove the filter, because the main frame does not think
> HTML-in-Canvas is enabled.
>
> Fix the cross-origin filter removal code to check for canvas subtree
> instead of using the feature flag. While this will remove filters in
> canvas subtrees even if HTML-in-Canvas is not enabled, it doesn't matter
> because the filters are not exposed to a11y and there is otherwise no
> way to see them.
>
> Bug: 550141694
> Change-Id: I3dc8f8d9f157ee2a3f73851687a04d537b1dcef9
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8310987
> Commit-Queue: Stephen Chenney <schenney@chromium.org>
> Reviewed-by: Stefan Zager <szager@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1689553}

(cherry picked from commit d514b24be816134378c44660782560b16685a29e)

Bug: 556632614,550141694
Change-Id: I3dc8f8d9f157ee2a3f73851687a04d537b1dcef9
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8354164
Reviewed-by: Stephen Chenney <schenney@chromium.org>
Reviewed-by: Stefan Zager <szager@chromium.org>
Commit-Queue: Stephen Chenney <schenney@chromium.org>
Cr-Commit-Position: refs/branch-heads/7977@{#2392}
Cr-Branched-From: b7fe14017379ddffae396d944fb8b59a5896c261-refs/heads/main@{#1669021}

diff --git a/third_party/blink/renderer/core/paint/paint_property_tree_builder.cc b/third_party/blink/renderer/core/paint/paint_property_tree_builder.cc
index 293cc6a3c361b32ed073a50fac0ef52a5d597b95..37bbb43f5e0e58caa606e193ca294f57eb76ed73 100644
--- a/third_party/blink/renderer/core/paint/paint_property_tree_builder.cc
+++ b/third_party/blink/renderer/core/paint/paint_property_tree_builder.cc
@@ -2533,9 +2533,7 @@ void FragmentPaintPropertyTreeBuilder::UpdateFilter() {
       EffectPaintPropertyNode::FilterInfo filter_info;
       UpdateFilterEffect(object_, properties_->Filter(), filter_info);
       bool is_filter_disallowed =
-          RuntimeEnabledFeatures::CanvasDrawElementEnabled(
-              object_.GetDocument().GetExecutionContext()) &&
-          object_.IsInCanvasSubtree() && filter_info.operations.OriginTainted();
+          state.is_in_canvas_subtree && filter_info.operations.OriginTainted();
       if (!(filter_info.operations.IsEmpty() || is_filter_disallowed)) {
         state.filter_info =
             std::make_unique<EffectPaintPropertyNode::FilterInfo>(
@@ -4293,9 +4291,7 @@ void FragmentPaintPropertyTreeBuilder::PopulateBackdropFilterIfNeeded(
   }
   if (!operations.IsEmpty()) {
     bool is_filter_disallowed =
-        RuntimeEnabledFeatures::CanvasDrawElementEnabled(
-            object_.GetDocument().GetExecutionContext()) &&
-        object_.IsInCanvasSubtree() && operations.OriginTainted();
+        state.is_in_canvas_subtree && operations.OriginTainted();
     if (!is_filter_disallowed) {
       state.backdrop_filter_info =
           base::WrapUnique(new EffectPaintPropertyNode::BackdropFilterInfo{
